Adds basic structure of Stable diffusion and it's usecases. Part of Unit 12 (#164)
johko/computer-vision-course: chapters/en/Unit 5 - Generative Models/Diffusion models/stable_diffusion.mdx

- How can we add-in good inductive biases? 

Since, we are trying to generate something new(e.g., a realistic Pokemon), we need a way to go beyond the images we have seen before(e.g., an anime Pokemon). That's where U-Net and self-attention come into the picture. Given a noisy version of an image, the model is tasked with predicting the denoised version based on additional clues such as a text description of the image. Ok, how do we actually feed this conditioning information into the U-Net for it to use as it makes predictions? The answer is something called cross-attention. Scattered throughout the U-Net are cross-attention layers. 
Each spatial location in the U-Net can 'attend' to different tokens in the text conditioning, bringing in relevant information from the prompt. 

## How to use `text-to-image`, `image-to-image`, Inpainting in Diffusers
This section introduces useful usecases and how can we use perform these tasks using Diffusers library.
- `text-to-image`
The idea is to pass in the text prompt, which is converted to the output image.

<iframe
	src="https://hysts-controlnet-v1-1.hf.space/"
	frameborder="0"
	width="850"
	height="450">
</iframe>

> 

Reviewer: ```suggestion
Let's install the `diffusers` library first.
```

> Using the `diffusers` library you can get `text-to-image` working in 2 steps.

Reviewer: Can you add diffusers installation here? 
```bash
pip install diffusers
```

```python
from diffusers import AutoPipelineForText2Image
import torch

pipeline = AutoPipelineForText2Image.from_pretrained(
	"runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, variant="fp16"
).to("cuda")
generator = torch.Generator("cuda").manual_seed(31)
image = pipeline("Astronaut in a jungle, cold color palette, muted colors, detailed, 8k", generator=generator).images[0]
image
```

- image-to-image
```python

> 

Reviewer: ```suggestion
```

# Step 1
import torch
from diffusers import AutoPipelineForImage2Image
from diffusers.utils import load_image, make_image_grid

pipeline = AutoPipelineForImage2Image.from_pretrained(
    "kandinsky-community/kandinsky-2-2-decoder", torch_dtype=torch.float16, use_safetensors=True
)
pipeline.enable_model_cpu_offload()
# remove following line if xFormers is not installed or you have PyTorch 2.0 or higher installed
pipeline.enable_xformers_memory_efficient_attention()

# Step 2

> 

Reviewer: ```suggestion
```

# Load an image to pass to the pipeline:
init_image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/cat.png")

#Step 3
# Pass a prompt and image to the pipeline to generate an image:
prompt = "cat wizard, gandalf, lord of the rings, detailed, fantasy, cute, adorable, Pixar, Disney, 8k"
image = pipeline(prompt, image=init_image).images[0]
make_image_grid([init_image, image], rows=1, cols=2)
```

- Inpainting
```python

> 

Reviewer: ```suggestion
```

# Step 1
# Load the pipeline
import torch
from diffusers import AutoPipelineForInpainting
from diffusers.utils import load_image, make_image_grid

pipeline = AutoPipelineForInpainting.from_pretrained(
    "kandinsky-community/kandinsky-2-2-decoder-inpaint", torch_dtype=torch.float16
)
pipeline.enable_model_cpu_offload()
# remove following line if xFormers is not installed or you have PyTorch 2.0 or higher installed
pipeline.enable_xformers_memory_efficient_attention()

# Step 2

# Load the base and mask images:
init_image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/inpaint.png")
mask_image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/inpaint_mask.png")

# Step 3

> 

Reviewer: ```suggestion
```

# Introduction to Stable Diffusion
This chapter introduces the building blocks of Stable Diffusion which is a generative artificial intelligence (generative AI) model that produces unique photorealistic images from text and image prompts. It originally launched in 2022 and was made possible thanks to a collaboration with

> [Stability AI](https://stability.ai/), [RunawayML](https://runwayml.com/) and CompVis Group at LMU Munich following the [paper](https://arxiv.org/pdf/2112.10752.pdf).

Reviewer 2: ```suggestion
[Stability AI](https://stability.ai/), [RunwayML](https://runwayml.com/) and CompVis Group at LMU Munich following the [paper](https://arxiv.org/pdf/2112.10752.pdf).
```

- How can we add-in good inductive biases? 

Since, we are trying to generate something new(e.g., a realistic Pokemon), we need a way to go beyond the images we have seen before(e.g., an anime Pokemon). That's where U-Net and self-attention come into the picture. Given a noisy version of an image, the model is tasked with predicting the denoised version based on additional clues such as a text description of the image. Ok, how do we actually feed this conditioning information into the U-Net for it to use as it makes predictions? The answer is something called cross-attention. Scattered throughout the U-Net are cross-attention layers. 
Each spatial location in the U-Net can 'attend' to different tokens in the text conditioning, bringing in relevant information from the prompt. 

> ## How to use `text-to-image`, `image-to-image`, Inpainting in Diffusers

Reviewer: ```suggestion
## How to use `text-to-image`, `image-to-image`, Inpainting Models in Diffusers
```

> This section introduces useful usecases and how can we use perform these tasks using Diffusers library.

Reviewer 2: ```suggestion
This section introduces helpful usecases and how we can perform these tasks using the [Diffusers](https://github.com/huggingface/diffusers) library.
```

> - `text-to-image`

Reviewer: ```suggestion
- Steps for `text-to-image` inference
```

The idea is to pass in the text prompt, which is converted to the output image.

<iframe
	src="https://hysts-controlnet-v1-1.hf.space/"
	frameborder="0"
	width="850"
	height="450">
</iframe>

Using the `diffusers` library you can get `text-to-image` working in 2 steps.

Let's install the `diffusers` library first.

> ```python

Reviewer: ```suggestion
```bash
```

pip install diffusers
```

```python
from diffusers import AutoPipelineForText2Image
import torch

pipeline = AutoPipelineForText2Image.from_pretrained(
	"runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, variant="fp16"
).to("cuda")
generator = torch.Generator("cuda").manual_seed(31)
image = pipeline("Astronaut in a jungle, cold color palette, muted colors, detailed, 8k", generator=generator).images[0]
image
```

> - Steps for image-to-image

Reviewer: ```suggestion
- Steps for `image-to-image` inference
```
